packages/dds/cell/src/test/cell.spec.ts

```diff
@@ -132,29 +133,83 @@ describe("Cell", () => {
+                cell.set("testValue");
 
-            containerRuntimeFactory.processAllMessages();
+                containerRuntimeFactory.processAllMessages();
 
-            assert.equal(cell.get(), "testValue", "Could not retrieve cell value");
-            assert.equal(cell2.get(), "testValue", "Could not retrieve cell value from remote client");
+                assert.equal(cell.get(), "testValue", "Could not retrieve cell value");
+                assert.equal(cell2.get(), "testValue", "Could not retrieve cell value from remote client");
 
-            cell.delete();
+                cell.delete();
 
-            containerRuntimeFactory.processAllMessages();
+                containerRuntimeFactory.processAllMessages();
 
-            assert.equal(cell.get(), undefined, "Could not delete cell value");
-            assert.equal(cell2.get(), undefined, "Could not delete cell value from remote client");
+                assert.equal(cell.get(), undefined, "Could not delete cell value");
+                assert.equal(cell2.get(), undefined, "Could not delete cell value from remote client");
+            });
+        });
+
+        describe("Garbage Collection", () => {
+            let subCellCount = 0;
+            let expectedRoutes: string[] = [];
+
+            beforeEach(() => {
+                subCellCount = 0;
+                expectedRoutes = [];
+            });
+
+            // Return the remote SharedCell because we want to verify its summary data.
+            const getSharedObject = () => cell2;
+
+            async function addOutboundRoutes() {
+                const factory = new CellFactory();
+                const newSubCell = factory.create(dataStoreRuntime, `subCell-${++subCellCount}`);
+                cell.set(newSubCell.handle);
+                expectedRoutes = [ newSubCell.handle.absolutePath ];
+                containerRuntimeFactory.processAllMessages();
+            }
+
+            async function deleteOutboundRoutes() {
+                cell.delete();
+                expectedRoutes = [];
+                containerRuntimeFactory.processAllMessages();
+            }
+
+            async function addNestedHandles() {
+                const factory = new CellFactory();
+                const newSubCell = factory.create(dataStoreRuntime, `subCell-${++subCellCount}`);
+                const newSubCell2 = factory.create(dataStoreRuntime, `subCell-${++subCellCount}`);
+                const containingObject = {
+                    subcellHandle: newSubCell.handle,
+                    nestedObj: {
+                        subcell2Handle: newSubCell2.handle,
+                    },
+                };
+                cell.set(containingObject);
+                expectedRoutes = [ newSubCell.handle.absolutePath, newSubCell2.handle.absolutePath ];
+                containerRuntimeFactory.processAllMessages();
+            }
+
+            const gcTestProvider: IGCTestProvider = {
```

Comment:
If you squeeze, you can see that all the code under describe() is basically a class (I'll call it GCCellTestProvider), followed by one-liner - runGCTests(new GCCellTestProvider()).
There is no need for beforeEach(), local variables (these are member variables of class).
IMHO class better represents the intention and relationships between methods and variables, otherwise one continuously needs to ask questions like "what's the scope/ lifetime of this variable?"

Reply:
Great suggestion! I changed it to a class.
I put the class definition in the describe block itself for each test. Should it be outside the describe blocks or perhaps in a separate file?

Reply:
I've looked at just one implementation, and asked question about cell2 lifetime management. I think if the intention for this class to be self-contained, then it's better to move it out to global scope, as this will ensure compiler will help you find dependencies on state that should not be there. But if the intention is to use some of the state setup above, then you can either leave it in place, or find other ways to pass it (like creating a factory function that returns new instance of class bound to some state, similar how mixins do it - see for example mixinRequestHandler.

Basically let's try to apply same quality to test code - how would you write similar shipping code to ensure that it's easier to find bugs and understand scoping / lifetime management of various pieces?
I find that our test code is quite often harder to read then shipping code :(

Otherwise looks much better IMHO!

Reply:
I thought about this too. I wanted to re-use the setup above. But I think I can move the creation of the DDS to a function and call it from the constuctor.
I agree about our test code not being readable! I will try to clean it up as much as I can here :-).

Reply:
Okay. So I made the class independent. Moved out creation of DDSs into a separate function which is called by the class and from the rest of the test code.
I cleaned up the tests as much as I could for now. I think they can be cleaned further but I would rather do that in a separate change.